crates/ruff_python_parser/src/string_token_flags.rs

```diff
@@ -0,0 +1,234 @@
+        } else {
+            QuoteStyle::Single
+        }
+    }
+
+    pub const fn is_triple_quoted(self) -> bool {
+        self.0.contains(StringFlagsInner::TRIPLE_QUOTED)
+    }
+
+    pub const fn quote_str(self) -> &'static str {
+        if self.is_triple_quoted() {
+            match self.quote_style() {
+                QuoteStyle::Single => "'''",
+                QuoteStyle::Double => r#"""""#,
+            }
+        } else {
+            match self.quote_style() {
+                QuoteStyle::Single => "'",
+                QuoteStyle::Double => "\"",
+            }
+        }
+    }
+
+    pub const fn prefix_str(self) -> &'static str {
+        if self.is_ustring() {
+            return "u";
+        }
+        if self.is_bytestring() {
+            if self.is_raw() {
+                return "rb";
+            }
+            return "b";
+        }
+        if self.is_fstring() {
+            if self.is_raw() {
+                return "rf";
+            }
+            return "f";
+        }
+        if self.is_raw() {
+            return "r";
+        }
+        ""
+    }
+
+    pub fn prefix_len(self) -> TextSize {
+        self.prefix_str().text_len()
+    }
+
+    pub fn quote_len(self) -> TextSize {
+        if self.is_triple_quoted() {
+            TextSize::from(3)
+        } else {
+            TextSize::from(1)
+        }
+    }
+
+    #[must_use]
+    pub fn with_double_quotes(mut self) -> Self {
+        self.0 |= StringFlagsInner::DOUBLE;
+        self
+    }
+
+    #[must_use]
+    pub fn with_triple_quotes(mut self) -> Self {
+        self.0 |= StringFlagsInner::TRIPLE_QUOTED;
+        self
+    }
+
+    pub fn with_u_prefix(mut self) -> Result<Self, &'static str> {
+        if self.0.intersects(DISALLOWED_U_STRING_PREFIXES) {
+            Err("U-strings cannot have any other prefixes set")
+        } else {
+            self.0 |= StringFlagsInner::U_PREFIX;
+            Ok(self)
+        }
+    }
+
+    pub fn with_b_prefix(mut self) -> Result<Self, &'static str> {
+        if self.0.intersects(DISALLOWED_B_STRING_PREFIXES) {
+            Err("Bytestrings cannot have the `u` or `f` prefix also set")
+        } else {
+            self.0 |= StringFlagsInner::B_PREFIX;
+            Ok(self)
+        }
+    }
+
+    pub fn with_f_prefix(mut self) -> Result<Self, &'static str> {
+        if self.0.intersects(DISALLOWED_F_STRING_PREFIXES) {
+            Err("F-strings cannot have the `b` or `u` prefix also set")
+        } else {
+            self.0 |= StringFlagsInner::F_PREFIX;
+            Ok(self)
+        }
+    }
+
+    pub fn with_r_prefix(mut self) -> Result<Self, &'static str> {
+        if self.0.intersects(DISALLOWED_R_STRING_PREFIXES) {
+            Err("Raw strings cannot have the `u` prefix also set")
+        } else {
+            self.0 |= StringFlagsInner::R_PREFIX;
+            Ok(self)
+        }
+    }
+}
+
+/// TODO: Use this enum in `crates/ruff_python_formatter/src/string/mod.rs`?
```

Comment:
Next time I'll just `@` you on the PR 😆 I mainly just wanted to have the discussion about whether it's worth "unifying" the various enums we have scattered across the codebase for describing the two quote styles. Maybe it's not -- maybe it's small and obvious enough that it's okay to just duplicate it across the various crates. If it _is_ worth having a single enum somewhere, I'm not really sure where it should go. It feels odd to me to have it live in the parser crate, since -- while it's a useful enumeration for the parser -- it's not really anything parser-specific. It's really something quite abstract and intrinsic to the language (which is why we have similar enums in the linter, the stylist and the formatter already).

FTR, here's the very-similar enums we already have in various places:

https://github.com/astral-sh/ruff/blob/461cdad53a1569d6b6dfa242734b140bcad1b7b7/crates/ruff_python_formatter/src/string/mod.rs#L234-L242

https://github.com/astral-sh/ruff/blob/461cdad53a1569d6b6dfa242734b140bcad1b7b7/crates/ruff_python_codegen/src/stylist.rs#L113-L119

https://github.com/astral-sh/ruff/blob/461cdad53a1569d6b6dfa242734b140bcad1b7b7/crates/ruff_linter/src/rules/flake8_quotes/settings.rs#L9-L17

https://github.com/astral-sh/ruff/blob/461cdad53a1569d6b6dfa242734b140bcad1b7b7/crates/ruff_python_literal/src/escape.rs#L1-L5

That last one is already quite nice and self-contained, so possibly I could use that one here -- but the `ruff_python_parser` crate doesn't currently depend on the `ruff_python_literal` crate. Is it worth adding a dependency between the two crates just for this? (Genuine question -- I'm curious for your opinion!)

Reply:
We probably need to move `string_token_flags` to the AST once you want to expose it on string nodes ;)

I would leave this for now and probably even remove the TODO. It's a nice cleanup we could do but having separate types also has its advantages.

Reply:
I definitely want to leave it for now! I tried, when preparing this PR, to see if I could unify this enum with the one in `stylist.rs`. It quickly unravelled into a "now you need to rewrite nearly every line in this crate" kind of thing, and I ran away 😆

I'll remove the TODO 👍